### PR TITLE
Geocoder city location lookup feature

### PR DESCRIPTION
**Pull request by TimCsaky**
# Description

supports ticket: [1090](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-1090)
- getting lat/long with the city lookup field
- add fields to location table
- modifying the backend data and transform service

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [ ] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/src/migrations/20200506163000-6-map-location-table.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            allowNull: false,
+            comment: 'Associated location id',
+            type: Sequelize.UUID,
+            references: {
+              model: 'location',
+              key: 'locationId'
+            }
+          },
+          latitude: {
+            allowNull: false,
```
Comment (loneil) [inline app/src/migrations/20200506163000-6-map-location-table.js:22]: I would consider not using varchars for this field. Best case for geography would be PostGIS but might not be important in the timeframe.
Could see what @parc-jason and @jujaga think too but there the point data type which might be perfect, or use appropriate precision decimal-type fields (since it's not nullable anyways).
Relevant: https://stackoverflow.com/a/8150944/324672

Comment (parc-jason) [reply app/src/migrations/20200506163000-6-map-location-table.js:22]: For now, let's just use [numeric data types](https://sequelize.org/master/class/lib/data-types.js~NUMBER.html).  Check over Postgres definitions [too](https://www.postgresql.org/docs/9.1/datatype-numeric.html).

Comment (TimCsaky) [reply app/src/migrations/20200506163000-6-map-location-table.js:22]: good call Lucas. PostGIS looks great, I'm not sure we have the geography module on our postgres image. point also sounds good. i'll look into these later

File: app/src/services/dataService.js [hunk 2]
```diff
@@ -249,12 +255,18 @@ module.exports = {
...
+
+        const locationObj = await db.Location.create({...location, ipcPlanId: ipcPlanId}, {transaction: t});
+        let locationId = locationObj.locationId;
 
         await db.Contact.bulkCreate(contacts.map(c => {
           return {...c, ipcPlanId: ipcPlanId};
         }), {transaction: t});
 
+        await db.MapLocation.bulkCreate(mapLocations.map(ml => {
+          return {...ml, locationId: locationId};
```
Comment (jujaga) [inline app/src/services/dataService.js:267]: As you only use locationId once in this function, you can streamline this directly into line 267 and drop the let locationId on line 260.
```js
return {...ml, locationId: locationObj.locationId};
```

File: app/src/services/dataService.js [hunk 3]
```diff
@@ -198,6 +201,9 @@ module.exports = {
           {
             model: db.Location,
           },
+          {
+            model: db.MapLocation,
+          },
```
Comment (TimCsaky) [inline app/src/services/dataService.js:206]: i think i'm doing this wrong. i've tried:
```
const ipcObj = await db.IPCPlan.findByPk(id,
        {
          order: [
            [db.InspectionStatus, 'createdAt', 'DESC']
          ],
          include: [
            {
              model: db.Business,
            },
            {
              model: db.Contact,
            },
            {
              model: db.Location,
              include: [
                {
                  model: db.MapLocation,
                },
              ]
            },
            {
              model: db.InspectionStatus
            },
            {
              model: db.Note
            }],

          rejectOnEmpty: true
        }
      );
````

File: app/src/models/location.js [hunk 4]
```diff
@@ -35,6 +35,16 @@ module.exports = (sequelize, DataTypes) => {
       type: DataTypes.STRING(255),
       unique: false
     },
+    cityLatitude: {
+      allowNull: false,
```
Comment (parc-jason) [inline app/src/models/location.js:39]: these will need to allow nulls.

Comment (TimCsaky) [reply app/src/models/location.js:39]: ah yes. just saw that. ty

File: app/src/services/dataService.js [hunk 5]
```diff
@@ -249,12 +249,14 @@ module.exports = {
...
 
         await db.Business.create({...business, ipcPlanId: ipcPlanId}, {transaction: t});
+
         await db.Location.create({...location, ipcPlanId: ipcPlanId}, {transaction: t});
 
         await db.Contact.bulkCreate(contacts.map(c => {
           return {...c, ipcPlanId: ipcPlanId};
         }), {transaction: t});
 
+
```
Comment (jujaga) [inline app/src/services/dataService.js:259]: Unnecessary whitespace added on lines 252 and 259

File: app/src/services/transformService.js [hunk 6]
```diff
@@ -40,6 +40,11 @@ const transformService = {
       const ipcPlan = {...obj.ipcPlan};
 
       const location = {...obj.location};
+
+      // split city lookup result into city and province
+      location.city = obj.location.city.substring(0, obj.location.city.lastIndexOf(', '));
+      location.province = obj.location.city.substring(obj.location.city.lastIndexOf(', ') + 2);
```
Comment (jujaga) [inline app/src/services/transformService.js:46]: This logic could be unstable as it has a lot of assumptions on how something is structured. Instead of using substrings, you can do something more succinct with the `str.split()` function and then `str.trim()` off the whitespace around it. For example...
```js
"one two, bc".split(',').map(x => x.trim());
```
Yields `[ 'one two', 'bc' ]`. You can assign to the correct variables easily from there.

File: app/frontend/src/components/form/Step2.vue [hunk 7]
```diff
@@ -604,6 +616,18 @@ export default {
...
+    cityLatitude: {
+      get() {
+        return this.location.cityLatitude;
+      },
+      set(value) { this.updateLocation({['cityLatitude']: value}); }
+    },
+    cityLongitude: {
+      get() {
+        return this.location.cityLongitude;
+      },
```
Comment (jujaga) [inline app/frontend/src/components/form/Step2.vue:628]: These gets don't need the contents in the curly braces to be expanded - can follow the style of the previous getters and setters.

File: app/src/services/transformService.js (hunk 6 above)
Comment (parc-jason) [reply app/src/services/transformService.js:46]: We have to be careful here.  Localities have their own comma (ex. Victor Rd, Mayne Island,  BC), so splitting on comma isn't a great idea.  But even going with lastIndex, we don't know that city has come from the lookup.  Users can enter whatever they want.  So, let's try to parse off ", BC" and get city/province, but if that fails, it all goes into city.  Also, now that I look at it, let's do nothing.  In Read only mode/word doc, you lose the ", BC".

Comment (TimCsaky) [reply app/src/services/transformService.js:46]: yes. good call!

Comment (jujaga) [reply app/src/services/transformService.js:46]: It's a fair point - there's too many possibilities for format - keeping what the user entered w/o transform is safer at this point.

File: app/frontend/src/store/modules/form.js [hunk 8]
```diff
@@ -161,6 +164,9 @@ export default {
     updateLocation: (state, obj) => {
       Object.assign(state.location, obj);
     },
+    updateMapLocation: (state, obj) => {
+      Object.assign(state.mapLocations[0], obj);
+    },
```
Comment (loneil) [inline app/frontend/src/store/modules/form.js:169]: What's this about? The coords are part of the location object aren't they?

File: app/frontend/src/store/modules/form.js [hunk 9]
```diff
@@ -196,7 +202,8 @@ export default {
           contacts: state.contacts,
           ipcPlan: state.ipcPlan,
           covidContact: state.covidContact,
-          location: state.location
+          location: state.location,
+          mapLocations: state.mapLocations
```
Comment (loneil) [inline app/frontend/src/store/modules/form.js:206]: Same as above
